lte/gateway/python/magma/pipelined/app/enforcement_stats.py

```diff
@@ -186,18 +213,26 @@ def _handle_error(self, ev):
         self._msg_hub.handle_error(ev)
 
     # pylint: disable=protected-access,unused-argument
-    def _get_rule_match_flow_msgs(self, imsi, _, __, ip_addr, ambr, rule, version):
+    def _get_rule_match_flow_msgs(self, imsi, _, __, ip_addr, ambr, rule, version, ng_session_id=0):
         """
         Returns flow add messages used for rule matching.
         """
         rule_num = self._rule_mapper.get_or_create_rule_num(rule.id)
         self.logger.debug(
             'Installing flow for %s with rule num %s (version %s)', imsi,
             rule_num, version)
-        inbound_rule_match = _generate_rule_match(imsi, ip_addr, rule_num,
-                                                  version, Direction.IN)
-        outbound_rule_match = _generate_rule_match(imsi, ip_addr, rule_num,
-                                                   version, Direction.OUT)
+        if ng_session_id:
+            inbound_rule_match = _generate_rule_match(imsi, ip_addr, rule_num,
+                                                      version, Direction.IN,
+                                                      ng_session_id)
+            outbound_rule_match = _generate_rule_match(imsi, ip_addr, rule_num,
+                                                       version, Direction.OUT,
+                                                       ng_session_id)
```

Comment:
why do we need rule_num and ng_session_id ?
I would like to see if we can have single ID to identify a rule from given flow.

Reply:
ng_session_id is required for distinguish between 4G and 5G.
As 5G design, apart from stats, the UPF will send the snapshoot of active session of session information periodically to SMF.
session_snapshot {
subscriber_id: "208950000000033"
session_version: 3
local_f_teid: 2147483654
}

Reply:
are we not sending ng_session_id back to sessionD?

Reply:
Yes, we are sending ng_session_id value periodically for each session.
Here, local_f_teid value is ng_session_id.

Reply:
why this name is changed ? is ng_session_id same as local_f_teid?

Reply:
We are used below grpc method for sending snapshoot of session information to SMF:
// Current Session snapshot in UPF.
message UPFSessionState {
  string subscriber_id = 1;       // Subscriber ID
  uint32 session_version = 2;      // Successful Version number snapshot from UPF. If its not the expected parse Failure report.
  uint32 local_f_teid = 3;        // Local TEID to uniquely identify the session
}
Same format we are receiving upf session requests from SMF
//SET message - SMF to Upf Session Requests
message SessionSet {
  string subscriber_id = 1;
  uint32 local_f_teid = 2;
  uint32 session_version = 3;
  NodeID node_id = 4;
  Fsm_state state = 5;
  repeated SetGroupPDR set_gr_pdr = 6;
}
We are set the value local_f_teid as ng_session_id (Just changed the name as a unique id for ng) for in rpc_servicer.py
ng_session_id = request.local_f_teid

Reply:
can you rename it ng_session_id to  local_f_teid_ng to avoid any confusion.

Reply:
Ok.I will do it now.

Reply:
can you also rebase these commits on latest master.

Reply:
Changes done. I did rebase it